pkg/query/querier_test.go

```diff
@@ -22,23 +22,180 @@ import (
-	queryableCreator := NewQueryableCreator(nil, testProxy, "test")
+type recordRequestsStoreServer struct {
+	// This field just exist to pseudo-implement the unused methods of the interface.
+	storepb.StoreServer
 
-	oneHourMillis := int64(1*time.Hour) / int64(time.Millisecond)
-	queryable := queryableCreator(false, oneHourMillis, false)
+	reqs []storepb.SeriesRequest
+}
 
-	q, err := queryable.Querier(context.Background(), 0, 42)
-	testutil.Ok(t, err)
-	defer func() { testutil.Ok(t, q.Close()) }()
+func (s *recordRequestsStoreServer) Series(r *storepb.SeriesRequest, _ storepb.Store_SeriesServer) error {
+	s.reqs = append(s.reqs, *r)
+	return nil
+}
+
+// timeParse is a minimal function to parse time.Time.
+func timeParse(sec string) time.Time {
+	fl, _ := strconv.ParseFloat(sec, 64)
+	s, ns := math.Modf(fl)
+	return time.Unix(int64(s), int64(ns*float64(time.Second)))
+}
 
-	querierActual, ok := q.(*querier)
+// TODO(bwplotka): Result is not expected. Investigate.
+func TestQuerier_PromQLToStoreAPI(t *testing.T) {
+	defer leaktest.CheckTimeout(t, 10*time.Second)()
+
+	const staleness = 5 * time.Minute
 
-	testutil.Assert(t, ok == true, "expected it to be a querier")
-	testutil.Assert(t, querierActual.maxResolutionMillis == oneHourMillis, "expected max source resolution to be 1 hour in milliseconds")
+	for _, tcase := range []struct {
+		query      string
+		start, end time.Duration // From 0 unix time.
+		step       time.Duration
 
+		expectedRequests []storepb.SeriesRequest
+	}{
+		{
+			query: "up{}",
+			end:   -1 * time.Millisecond,
+
+			expectedRequests: []storepb.SeriesRequest{
+				{
+					MaxTime: timePast(-1 * time.Millisecond),
+					MinTime: timePast(-1*time.Millisecond - staleness),
+					Matchers: []storepb.LabelMatcher{
+						{
+							Type:  0,
+							Name:  "__name__",
+							Value: "up",
+						},
+					},
+					Aggregates: []storepb.Aggr{storepb.Aggr_COUNT, storepb.Aggr_SUM},
+				},
+			},
+		},
+		{
+			query: "up{}",
+			start: -24 * time.Hour,
+			end:   -1 * time.Millisecond,
+			step:  100 * time.Millisecond,
+
+			expectedRequests: []storepb.SeriesRequest{
+				{
+					MaxTime: timePast(-1 * time.Millisecond),
+					MinTime: timePast(-24*time.Hour - staleness),
+					Matchers: []storepb.LabelMatcher{
+						{
+							Type:  0,
+							Name:  "__name__",
+							Value: "up",
+						},
+					},
+					MaxResolutionWindow: 20, // Milis.
+					Aggregates:          []storepb.Aggr{storepb.Aggr_COUNT, storepb.Aggr_SUM},
+				},
+			},
+		},
+		{
+			query: "up[5m]",
+			end:   -1 * time.Millisecond,
+
+			expectedRequests: []storepb.SeriesRequest{
+				{
+					MaxTime: timePast(-1 * time.Millisecond),
+					MinTime: timePast(-1*time.Millisecond - staleness),
+					Matchers: []storepb.LabelMatcher{
+						{
+							Type:  0,
+							Name:  "__name__",
+							Value: "up",
+						},
+					},
+					Aggregates: []storepb.Aggr{storepb.Aggr_COUNT, storepb.Aggr_SUM},
+				},
+			},
+		},
+		{
+			query: "rate(up[5m:200s])",
```

Comment:
This does not work as expected (in my mind). Step is still passed as 100 ms no matter what. cc @brian-brazil for help =D

Reply:
Based my knowledge on: https://prometheus.io/blog/2019/01/28/subquery-support/#subqueries

Reply:
The step is the one for query_range. We're not passing any hints relating to subqueries currently.

Reply:
Thanks! Will look deeper in PromQL if it's sane to actually propagate that hint & how.